### PR TITLE
[rllib] Support training on multi-agent environments with shared rewards

### PR DESCRIPTION
**Pull request by eugenevinitsky**
## What do these changes do?

The environment is now wrapped in a (poorly named) RayEnv that, in the event the observation space is a list corresponding to a multiagent case, flattens all the lists into one long vector. The class also provides methods that cut up the vector into its original shape. 
There is now a loss called mProximalPolicyLoss. In the event of multiple agents, the runner calls it. It takes the flattened vector and cuts it up to form the appropriate loss. 
The advantages are duplicated for all agents, meaning that this currently only works for shared rewards. 
The changes did not interfere with other examples, as both CartPoleEnv and Pong still ran.

### CODE REVIEW

**Comment by robertnishihara** [comment]
ok to test

**Comment by robertnishihara** [comment]
Do you have an example of how to use `RayEnv`?

File: python/ray/rllib/__init__.py [hunk 1]
```diff
@@ -6,7 +6,6 @@
 from ray.rllib import ppo, es, dqn, a3c
 from ray.rllib.agent import _MockAgent, _SigmoidFakeData
 
-
```
Comment (richardliaw) [inline python/ray/rllib/__init__.py]: nit: need to leave this here to pass linting test

File: python/ray/rllib/envs.py [hunk 2]
```diff
@@ -5,8 +5,11 @@
 import gym
```
Comment (richardliaw) [inline python/ray/rllib/envs.py]: We actually just removed this file; It makes sense to rename this as vectorized_env or something along those lines.

File: python/ray/rllib/models/catalog.py [hunk 3]
```diff
@@ -54,6 +54,21 @@ def get_action_dist(action_space, dist_type=None):
                 return Deterministic, action_space.shape[0]
         elif isinstance(action_space, gym.spaces.Discrete):
             return Categorical, action_space.n
+        elif isinstance(action_space, list):
```
Comment (richardliaw) [inline python/ray/rllib/models/catalog.py]: Does it make sense to push this down to the Catalog level? I think it would be reasonable if this were done as a helper outside of this function.

**Comment by ericl** [comment]
Thanks for contributing this!

My main question is how we can get the multiagent support to work with all Rllib algorithms -- currently it seems like you need to modify the PPO loss and runner specifically. I'll also think about this more over the weekend.

File: python/ray/rllib/models/catalog.py (hunk 3 above)
Comment (cathywu) [reply python/ray/rllib/models/catalog.py]: Could you explain what you mean by Catalog level?

Comment (ericl) [reply python/ray/rllib/models/catalog.py]: I'm not sure what Catalog level means, but it seems like the "distribution list" here can be kept inside an ActionDistribution class (e.g. if you call sample() internally it would sample from N distributions and concat them). That way, you wouldn't need to special case handling of this distribution since it would behave like a normal action distribution. Let me know if that makes sense.

File: python/ray/rllib/models/catalog.py [hunk 4]
```diff
@@ -121,7 +136,11 @@ def get_preprocessor(cls, env, options=dict()):
             env.observation_space.shape = ()
 
         env_name = env.spec.id
-        obs_shape = env.observation_space.shape
+
+        if isinstance(env.observation_space, list):
+            obs_shape = env.observation_space[0].shape
+        else:
+            obs_shape = env.observation_space.shape
```
Comment (ericl) [inline python/ray/rllib/models/catalog.py]: What if we did np.stack in the RayEnv / preprocessor so that the shape becomes a tensor like
`[len(original_space), original_space[0]]`?

File: python/ray/rllib/ppo/loss.py [hunk 5]
```diff
@@ -93,3 +94,115 @@ def compute(self, observation):
...
+        self.entropy = []
+        self.mean_entropy = []
+        self.surr1 = []
+        self.surr2 = []
+        self.surr = []
+        self.mean_policy_loss = []
+        for i in range(len(act_list)):
+            config["model_num"] = str(i)
+            self.prev_dist.append(distribution_class(prev_logits[i]))
+            self.curr_logits.append(ModelCatalog.get_model(
```
Comment (ericl) [inline python/ray/rllib/ppo/loss.py]: I think you can actually use the original policy loss and instead use a custom model that can interpret the N-agent observations of shape `[N, original_shape]` and output `[N, action_space]`. GAE might work with no special handling then.

File: python/ray/rllib/envs.py [hunk 6]
```diff
@@ -5,8 +5,11 @@
...
 import logging
 import time
+import numpy as np
+import tensorflow as tf
 
 from ray.rllib.models import ModelCatalog
+from ray.rllib.utils.overrides import overrides
```
Comment (ericl) [inline python/ray/rllib/envs.py]: Hm you don't seem to need this, so it would be nice to get rid of the extra dependency.

File: python/ray/rllib/models/catalog.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/rllib/models/catalog.py]: Oh I meant "Catalog level' as in modifying the Catalog code; alternative, you could have the Evaluator take care of this by something along the lines of `list_of_dists =
 [ModelCatalog.get_action_dist(space) for space in vectorized_action_space`

Comment (ericl) [reply python/ray/rllib/models/catalog.py]: Yeah the issue there is that now you have to special case the action distribution being a list everywhere.

File: python/ray/rllib/models/catalog.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/models/catalog.py]: Hm maybe the right thing to do is to support tuple spaces in RLlib in general. We need this for Carla to work anyways.

Comment (ericl) [reply python/ray/rllib/models/catalog.py]: Nevermind, the above probably won't work if the spaces are irregularly shaped.

File: python/ray/rllib/ppo/loss.py (hunk 5 above)
Comment (eugenevinitsky) [reply python/ray/rllib/ppo/loss.py]: @ericl Sorry, was on break all week. This seems like a sensible approach, and we can just extend GAE later if we need non-shared rewards?

File: python/ray/rllib/models/catalog.py (hunk 3 above)
Comment (eugenevinitsky) [reply python/ray/rllib/models/catalog.py]: Makes sense, I'll do it this way.

File: python/ray/rllib/ppo/loss.py (hunk 5 above)
Comment (eugenevinitsky) [reply python/ray/rllib/ppo/loss.py]: As an update, this seems to be working. I'll push up a cleaned up version soon.

Comment (ericl) [reply python/ray/rllib/ppo/loss.py]: Awesome, looking forward to it.

For the non-shared reward case: yeah, we'd need something like that, which could be tricky to support in a more general way. There seem to be a few types of multi-agent: (shared reward, shared model), (shared reward, non-shared model), (non-shared reward, shared model), (non-shared-reward, non-shared model).

Seems like this PR can cover the first two cases. The third you could possibly handle by splitting the observations up so each agent's experience is handled separately. The fourth is the tricky one, the only easy thing that comes to mind is Philipp's suggestion to train one type of agent a time, so it can be treated as the third case.

Comment (eugenevinitsky) [reply python/ray/rllib/ppo/loss.py]: @ericl 
A quick question, I'm currently doing the splitting of the inputs and outputs in a new multiagent model; I'm passing in a set of lists indicating how the splits should be done. I'm doing this because the model doesn't have access to the original environment observation and action space. This doesn't seem like a good way, do you have any suggestions?

Comment (ericl) [reply python/ray/rllib/ppo/loss.py]: Yeah, in the Carla env code, we make the original space dimensions a configuration parameter in catalog.py. You can do the same (make the original obs spaces a required configuration parameter of the model).

There might be a nicer way but this should be fine as long as we document it well.

Comment (eugenevinitsky) [reply python/ray/rllib/ppo/loss.py]: Submitted a new pull request that I think is a much cleaner approach so I'll close this one then?

**Comment by ericl** [comment]
Superceded by https://github.com/ray-project/ray/pull/1416
